stdlib/src/python/_cpython.mojo

```diff
@@ -1393,6 +1420,31 @@ struct CPython:
         self._inc_total_rc()
         return r
 
+    fn PySlice_FromSlice(inout self, slice: Slice) -> PyObjectPtr:
+        # Convert Mojo Slice to Python slice parameters
+        # Note: Deliberately avoid using `span.indices()` here and instead pass
+        # the Slice parameters directly to Python. Python's C implementation
+        # already handles such conditions, allowing Python to apply its own slice
+        # handling.
+        var py_start = self.Py_None()
+        var py_stop = self.Py_None()
+        var py_step = self.PyLong_FromLong(slice.step)
+
+        if slice.start:
+            py_start = self.PyLong_FromLong(slice.start.value())
+        if slice.end:
+            py_stop = self.PyLong_FromLong(slice.end.value())
+
+        var py_slice = self.PySlice_New(py_start, py_stop, py_step)
+
+        if py_start != self.Py_None():
+            self.Py_DecRef(py_start)
+        if py_stop != self.Py_None():
+            self.Py_DecRef(py_stop)
+        self.Py_DecRef(py_step)
+
+        return py_slice
+
```

Comment:
**Suggestion** This file (`_cpython.mojo`) is mostly intended to be low level wrappers around functions in the CPython CAPI. Since `PySlice_FromSlice` isn't a CPython function, I think the logic from this method might fit better as an initializer on `PythonObject`, e.g.:

```mojo
struct PythonObject(..):
    # ...

    fn __init__(inout self, slice: Slice):
        ...
```

Additionally, instead of calling low-level methods like `PyLong_FromLong()` or `Py_None()` directly, the code might be improved by using the existing higher level safe wrappers for those functions (provided by `PythonObject` as well). So the final code might look something like:

```mojo
    fn __init__(inout self, slice: Slice) -> PyObjectPtr:
        # Convert Mojo Slice to Python slice parameters
        # Note: Deliberately avoid using `span.indices()` here and instead pass
        # the Slice parameters directly to Python. Python's C implementation
        # already handles such conditions, allowing Python to apply its own slice
        # handling.
        var py_start = PythonObject(None)
        var py_stop = PythonObject(None)
        var py_step = PythonObject(slice.step)

        if slice.start:
            py_start = PythonObject(slice.start.value())
        if slice.end:
            py_stop = PythonObject(slice.end.value())

        var py_slice = cpython.PySlice_New(py_start, py_stop, py_step)

        return PythonObject(py_slice)
```

This has the advantage of allowing us to omit the manual ref count management logic as well 🙂 

Let me know what you think, or if there's a subtlety I missed about how you're using that means this might not be feasible for your use-case.

-----

For context, very roughly speaking, there are currently three files in the stdlib pertaining to interop, roughly in order from low-level to high-level:

* `_cpython.mojo` — low-level, "raw" bindings to the CPython API. In some future world, these might even become auto-generated bindings. They should have minimal logic.

* `python_object.mojo` — defines `PythonObject`, which interfaces between Mojo types and valid Python object values.

* `python.mojo` — high-level, safe wrappers around functionality from the CPython API.

Reply:
Hi, thank you for the time and feedback! :) 

> Let me know what you think, or if there's a subtlety I missed about how you're using that means this might not be feasible for your use-case.

I originally wanted to avoid `PythonObject` within `__getitem__`, as i noticed marginal gains. This was the intent of `PySlice_FromSlice`, to construct a `PyObjectPtr`. However, I overlooked the fact that the other `From*` functions are actually part of the C API.

I agree with your point about maintaining the minimalism of `_cpython.mojo`. What do you think of refactoring `PySlice_FromSlice` not inline of `__init__` but rather as a helper within `python_object.mojo`/`python.mojo`? This way we can maintain `PyObjectPtr` within `__getitem__`. 

> Additionally, instead of calling low-level methods like PyLong_FromLong() or Py_None() directly, the code might be improved by using the existing higher level safe wrappers for those functions (provided by PythonObject as well). So the final code might look something like:

On this note, I noticed what seems to be lifetime issues. To apply this suggestion, I would have to do something like the following:
```mojo
  ...
        
  cpython = _get_global_python_itf().cpython()
  var py_slice = cpython.PySlice_New(
    py_start.unsafe_as_py_object_ptr(),
    py_stop.unsafe_as_py_object_ptr(),
    py_step.unsafe_as_py_object_ptr()
  )
    
  _ = py_start^    
  _ = py_stop^    
  _ = py_step^    
  
  return py_slice
```

This might be worth exploring, following 400866c0abc2d0deba91645a53a86242e9028709